---
title: "Annotation"
format: html
editor: visual
---

1.  Cell clustering

```{r}

TTs.combined.sct <- RunUMAP(TTs.combined.sct, reduction = "pca", dims = 1:30)
TTs.combined.sct <- FindNeighbors(TTs.combined.sct, reduction = "pca", dims = 1:30)

# Run FindClusters using three different resolutions
TTs.combined.sct <- FindClusters(TTs.combined.sct, resolution = 0.3, algorithm = 2)
TTs.combined.sct <- FindClusters(TTs.combined.sct, resolution = 0.9, algorithm = 2)
TTs.combined.sct <- FindClusters(TTs.combined.sct, resolution = 0.5, algorithm = 2)

# check the cluster location
DimPlot(TTs.combined.sct, label=TRUE, group.by = "integrated_snn_res.0.9")

# check the cluster location
DimPlot(TTs.combined.sct, label=TRUE,  group.by = "integrated_snn_res.0.3")

# check the cluster location
DimPlot(TTs.combined.sct, label=TRUE,  group.by = "integrated_snn_res.0.5")
```

Run clustree

```{r}
library(clustree)
clustree(TTs.combined.sct, prefix = "integrated_snn_res.")

Idents(TTs.combined.sct) <- "integrated_snn_res.0.5"
```

```{r}
library(SingleR)
library(SingleCellExperiment)
```

1.  Load reference dataset

```{r}
# Load the dataset
ref_dataset <- readRDS("/home/people/s233658/projects/applied_single_cell/data/9dff3651-e629-4519-aaab-dbd21b6b02b1.rds")
```

Convert cell_type to character

```{r}
#Convert cell_type to character
ref_dataset$cell_type <- as.character(ref_dataset@meta.data$cell_type)
```

Subsetting by breast tissue

```{r}
# Subset Seurat object by tissue
ref_dataset_breast <- subset(ref_dataset, subset = tissue == "breast")
```

```{r}
dim(ref_dataset_breast)
```

Tranform reference dataset into singlecellexperiment object

```{r}
sce <- as.SingleCellExperiment(ref_dataset)
```

2.  Run SingleR

```{r}
integrated.seu <- readRDS("data/scRNA/TTs.combined.sct")
```

```{r}
# Run SingleR for annotation
annotations <- SingleR(
  test = integrated.seu,
  ref = sce,
  labels = ref_dataset$cell_type,
  de.method = 'wilcox' 
)
```

Inspect quality predictions

```{r}
# Inspect the resulting annotations
head(annotations)
# Inspect quality of the predictions
plotScoreHeatmap(ct_ann)
plotDeltaDistribution(ct_ann, ncol = 4, dots.on.top = FALSE)
```

3.  Add SingleR predictions to Seurat object

```{r}
# Add annotations to Seurat metadata
if (inherits(query_dataset, "Seurat")) {
  query_dataset[["SingleR.labels"]] <- annotations$labels
}
```

```{r}
# Seurat DimPlot with annotations
DimPlot(query_dataset, group.by = "SingleR.labels", label = TRUE, repel = TRUE)

```

```{r}
saveRDS(query_dataset, "/path/to/annotated_query_dataset.rds")
```

```{r}
library(Seurat)
library(Signac)
library(ggplot2)
```

```{r}
atac.integrated <- readRDS('data/scATAC/atac.integrated.rds')
rna.integrated <- readRDS('data/scRNA/rna.integrated.rds')
```

```{r}
rna.integrated <- NormalizeData(rna.integrated)
rna.integrated <- FindVariableFeatures(rna.integrated)
rna.integrated <- ScaleData(rna.integrated)
rna.integrated <- RunPCA(rna.integrated)
rna.integrated <- RunUMAP(rna.integrated, dims = 1:30)

atac.integrated <- RunTFIDF(atac.integrated)
atac.integrated <- FindTopFeatures(atac.integrated, min.cutoff = "q0")
atac.integrated <- RunSVD(atac.integrated)
atac.integrated <- RunUMAP(atac.integrated, reduction = "lsi", dims = 2:30, reduction.name = "umap.atac", reduction.key = "atacUMAP_")

```

## Identifying anchors between scRNA-seq and scATAC-seq datasets

```{r}
# quantify gene activity
gene.activities <- GeneActivity(atac.integrated, features = VariableFeatures(rna.integrated))

```
